Update `LightningDataModule` docs (#10678)
PyTorchLightning/pytorch-lightning: docs/source/extensions/datamodules.rst

        def test_dataloader(self):
            return DataLoader(self.mnist_test, batch_size=32)

        def predict_dataloader(self):
            return DataLoader(self.mnist_predict, batch_size=32)

---------------

LightningDataModule API
-----------------------
To define a DataModule the following methods are used to create train/val/test/predict dataloaders:

- :ref:`prepare_data<extensions/datamodules:prepare_data>` (how to download, tokenize, etc...)
- :ref:`setup<extensions/datamodules:setup>` (how to split, define dataset, etc...)
- :ref:`train_dataloader<extensions/datamodules:train_dataloader>`
- :ref:`val_dataloader<extensions/datamodules:val_dataloader>`
- :ref:`test_dataloader<extensions/datamodules:test_dataloader>`
- :ref:`predict_dataloader<extensions/datamodules:predict_dataloader>`

prepare_data
~~~~~~~~~~~~

> Use :meth:`~pytorch_lightning.core.hooks.DataHooks.prepare_data` method to do things that might write to disk or that need to be done only from a single process in distributed

Reviewer: ```suggestion
Downloading and saving data with multiple processes (distributed settings) will result in corrupted data.  Lightning ensures the :meth:`~pytorch_lightning.core.hooks.DataHooks.prepare_data` is called only within a single process, so you can safely add your downloading logic within.
```

            MNIST(os.getcwd(), train=False, download=True, transform=transforms.ToTensor())

> .. warning:: ``prepare_data`` is called from a single process across all nodes. It is not recommended to assign state here (e.g. ``self.x = y``).

Author: can someone confirm this?

Reviewer 2: Depends on the value of `LightningModule.prepare_data_per_node`:

https://github.com/PyTorchLightning/pytorch-lightning/blob/2036dfb5df2a48056d1c043ee311aed793187070/pytorch_lightning/trainer/connectors/data_connector.py#L159-L160

            if stage in (None, "fit"):
                mnist_full = MNIST(self.data_dir, train=True, download=True, transform=self.transform)
                self.mnist_train, self.mnist_val = random_split(mnist_full, [55000, 5000])

            # Assign Test split(s) for use in Dataloaders
            if stage in (None, "test"):
                self.mnist_test = MNIST(self.data_dir, train=False, download=True, transform=self.transform)

This method expects a ``stage`` argument.
It is used to separate setup logic for ``trainer.{fit,validate,test,predict}``. If ``setup`` is called with ``stage=None``,
we assume all stages have been set-up.

.. note:: :ref:`setup<extensions/datamodules:setup>` is called from every process across all the nodes. Setting state here is recommended.
.. note:: :ref:`teardown<extensions/datamodules:teardown>` can be used to clean up the state. It is also called from every process across all the nodes.

train_dataloader
^^^^^^^^^^^^^^^^

> Use the :meth:`~pytorch_lightning.core.hooks.DataHooks.train_dataloader` method to generate the train dataloader(s).

Reviewer 3: ```suggestion
Use the :meth:`~pytorch_lightning.core.hooks.DataHooks.train_dataloader` method to generate the training dataloader(s).
```

val_dataloader
^^^^^^^^^^^^^^

> Use the :meth:`~pytorch_lightning.core.hooks.DataHooks.val_dataloader` method to generate the val dataloader(s).

Reviewer 3: ```suggestion
Use the :meth:`~pytorch_lightning.core.hooks.DataHooks.val_dataloader` method to generate the validation dataloader(s).
```

            return DataLoader(self.mnist_val, batch_size=64)

> test_dataloader

Author: ```suggestion
.. _datamodule-test-dataloader-label:

test_dataloader
```

predict_dataloader
^^^^^^^^^^^^^^^^^^

> Use the :meth:`~pytorch_lightning.core.hooks.DataHooks.predict_dataloader` method to generate the predict dataloader(s).

Reviewer 3: ```suggestion
Use the :meth:`~pytorch_lightning.core.hooks.DataHooks.predict_dataloader` method to generate the prediction dataloader(s).
```

    class MNISTDataModule(pl.LightningDataModule):
        def predict_dataloader(self):
            return DataLoader(self.mnist_test, batch_size=64)
            return DataLoader(self.mnist_predict, batch_size=64)

transfer_batch_to_device
^^^^^^^^^^^^^^^^^^^^^^^^

> Override to define how you want to move an arbitrary batch to a device.

Reviewer 3: I preferred the previous way of separating API docs from user guide.

Author: won't this one be better? user will get the API arguments + the example was just repeated here and already present in the API docs.

    :noindex:

on_before_batch_transfer
^^^^^^^^^^^^^^^^^^^^^^^^

.. automethod:: pytorch_lightning.core.datamodule.LightningDataModule.on_before_batch_transfer
    :noindex:

on_after_batch_transfer
^^^^^^^^^^^^^^^^^^^^^^^

.. automethod:: pytorch_lightning.core.datamodule.LightningDataModule.on_after_batch_transfer
    :noindex:

on_load_checkpoint
~~~~~~~~~~~~~~~~~~

> 

Reviewer 3: from here on, the title levels are inconsistent, causing the index to be confused:

![image](https://user-images.githubusercontent.com/5495193/143036244-2fd9d085-7fe3-4e3c-8549-c473eb988c3f.png)

Not sure why.

Author: fixed!

.. automethod:: pytorch_lightning.core.datamodule.LightningDataModule.on_load_checkpoint
    :noindex:

on_save_checkpoint
~~~~~~~~~~~~~~~~~~

.. automethod:: pytorch_lightning.core.datamodule.LightningDataModule.on_save_checkpoint
    :noindex:

on_train_dataloader
~~~~~~~~~~~~~~~~~~~

.. automethod:: pytorch_lightning.core.datamodule.LightningDataModule.on_train_dataloader
    :noindex:

on_val_dataloader
~~~~~~~~~~~~~~~~~

.. automethod:: pytorch_lightning.core.datamodule.LightningDataModule.on_val_dataloader
    :noindex:

on_test_dataloader
~~~~~~~~~~~~~~~~~~

.. automethod:: pytorch_lightning.core.datamodule.LightningDataModule.on_test_dataloader
    :noindex:

on_predict_dataloader
~~~~~~~~~~~~~~~~~~~~~

.. automethod:: pytorch_lightning.core.datamodule.LightningDataModule.on_predict_dataloader
    :noindex:

teardown

> ~~~~~~~~~~~~~~~~~~~~~

Reviewer 3: too long underline
